### PR TITLE
[JENKINS-57516] - Support tag creation on tag push

### PR DESCRIPTION
**Pull request by sheldonrong**
Able to detect a tag creation from the Bitbucket PushHead Event and automatically create tags in Jenkins. This allows the workflow to build projects when tagging and/or pushing a commit.

This changeset adds a tag detection branch logic in the **NativeServerPushHookProcessor.HeadEvent**'s **heads** function.

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

**Comment by sheldonrong** [comment]
Setup Bitbucket server and jenkins server using docker container, and is able to verify that the tags can be instantly created once dev ran `git push --tags` using their git command line inside a repo.

**Comment by casz** [comment]
https://github.com/jenkinsci/bitbucket-branch-source-plugin/commit/cefd3175a10da2d413cf72e67a2a5342cf130ad4.patch

@sheldonrong
You might want to set up an email alias on your GitHub account 🥇 
Or look into fixing your git user.email config 😅 
to get proper credit for your commits 🏆 
https://help.github.com/articles/adding-an-email-address-to-your-github-account/
https://help.github.com/articles/setting-your-commit-email-address-in-git
then to amend your commits: https://stackoverflow.com/a/3042512

**Comment by sheldonrong** [comment]
@casz Thanks for reminding, updated the commit, and should be good now :)

**Comment by sheldonrong** [comment]
@casz looks like the **_continuous-integration/jenkins/pr-merge_** is stuck :)

**Comment by sheldonrong** [comment]
@casz any updates on this one?

**Comment by mortenkri** [comment]
Great if someone could push this forward. To use this is the only feature missing for us in our current CI/CD pipeline :)

**Comment by mortenkri** [comment]
Thx Joseph. Will test it as soon as I can :)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPushHookProcessor.java [hunk 1]
```diff
@@ -137,23 +138,29 @@ protected BitbucketServerRepository getRepository() {
             return result;
         }
 
-        private void addBranches(BitbucketSCMSource src, Map<SCMHead, SCMRevision> result) {
+        private void addBranchesAndTags(BitbucketSCMSource src, Map<SCMHead, SCMRevision> result) {
```
Comment (darxriggs) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPushHookProcessor.java]: A test covering this new logic would be more than welcome.

**Comment by mortenkri** [comment]
> Thx Joseph. Will test it as soon as I can :)

Verified that this works. Thanks a lot. Hopefully I can provide a PR for tag delete events :)

**Comment by ruhrohraggy** [comment]
Hoping someone can see this and respond -- it seems like this fixed an issue where once the Webhook was setup in Bitbucket to notify on Tag Created events, this was needed to process the event notification... Am I missing something, or is there still not logic to automatically include `Tag Created` as an option when automatically registering the webhook when the job is created...

**Pull request merged**